docs: add explanation about creating a slice (#147)
canonical/rockcraft: docs/how-to/create-slice.rst

.. code-block:: yaml

    <name_of_the_slice>
        <essential>
        <contents>
        <mutate>

| The essential part is optional and permits you to install the dependencies of the packages you want to slice.
| The contents part is required and permits you to keep the files from the package you want to keep in the slice.
| The mutate part is optional and allows the reproduction of maintainer scripts. (use this https://github.com/google/starlark-go)

Find dependencies of our packages
---------------------------------

So, the first step is to find the dependencies of the openssl package. We can easily do it with this command: 

.. code-block:: sh

    $ apt show openssl

.. code-block:: sh

    package: openssl
    Version: 3.0.2-0ubuntu1.7
    Origin: Ubuntu
    Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
    Original-Maintainer: Debian OpenSSL Team <pkg-openssl-devel@alioth-lists.debian.net>
    Bugs: https://bugs.launchpad.net/ubuntu/+filebug
    Depends: libc6 (>= 2.34), libssl3 (>= 3.0.2-0ubuntu1.2)

So we see that openssl depends on libc6 and libssl3 (these slices already exist; we will see them after)

Creating our slice
------------------

> Now we will create our openssl slice, but first, you need to clone this repository (our slice database):

Reviewer: ```suggestion
You now have everything needed to actually define the OpenSSL slice that will install the content you're looking to have in the final ROCK.

Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice _**bins**_? Let's go for it:

1. **What is the name of your slice definitions file?** It is a YAML file called ``openssl.yaml``
2. **What package name should be defined inside this file?** The package name is ``openssl``
3. **What is your slice name?** It should be called ``bins``
4. **What contents do you need from the OpenSSL package?** Just the binaries - ``/usr/bin/c_rehash`` and ``/usr/bin/openssl``
5. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``
    - **Do these two packages have slice definitions files upstream?** Yes, there's already a slice definitions file for [libc6](https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml) and another for [libssl3](https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml)
        - If these dependencies were not present in the upstream Chisel release, you'd also need to create their corresponding slice definitions
    - **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` itself.
    
With all this figured out, you can now create a new YAML file named ``openssl.yaml``, with the following content:

.. code-block:: yaml
    package: openssl
    slices:
        bins:
            essential:
            - libc6_libs
            - libc6_config
            - libssl3_libs
            - openssl_config
            contents:
              /usr/bin/c_rehash:
              /usr/bin/openssl:
            
        config:
            contents:
              /etc/ssl/private/:
              /etc/ssl/openssl.cnf:
              /usr/lib/ssl/certs:
              /usr/lib/ssl/openssl.cnf:
              /usr/lib/ssl/private:

Notice the unforeseen new slice ``config``. Because your OpenSSL binaries depend on the OpenSSL configuration files, and those weren't yet present anywhere in the Chisel releases upstream, you also need to create that slice! You may also ask _"why not put those configuration files inside the ``bins`` slice"_? You could! But we recommend, as a best practice, to separate and group contents according to their nature, as you may tomorrow need to create a new slice definition that only needs the OpenSSL configurations and not the binaries.

---

And that's it. This is your brand new slice definitions file, which will allow Chisel to install **just** the OpenSSL binaries (and their dependencies) into your ROCK! To learn about how to actually use this new slice definition file and publish it upstream for others to use, please check the following guides.
        
```

> .. code-block:: sh

Reviewer: can you pls remove everything from here until the end of the file? the content is already covered in the suggested commit above

**Let's assume you are trying to create a slice definition for installing the OpenSSL binary into your ROCK!**

Make sure the slice definition doesn't exist yet
-------------------------------------------------------------
You obviously want to avoid re-doing existing work, so before even considering creating your own slice definition, please make sure there isn't already something fitting of your needs. **How** ?

1. look into the upstream `chisel-releases <https://github.com/canonical/chisel-releases/tree/ubuntu-22.04/slices>`_ repository,
2. switch to the branch corresponding to the desired Ubuntu release for your ROCK,
3. scroll through the list of slice definition files, looking for your package name
    - if you find it, open it and try to find a slice name containing the bits and pieces you need from that package

> Structure of a slice definition file

Reviewer: ```suggestion
Structure of a slice definitions file
```

------------------------------------

There is **one slice definition file** for each Ubuntu package. All of the slice definition files follow the same structure:

.. code-block:: yaml

>     # (req) Name of the package. The slice definition file should be named accordingly (eg. "openssl.yaml")

Reviewer: ```suggestion
    # (req) Name of the package. 
    # The slice definition file should be named accordingly (eg. "openssl.yaml")
```

    package: <package-name>

    # (req) List of slices
    slices:

        # (req) Name of the slice
        <slice-name>:

            # (opt) Optional list of slices that this slice depends on
            essential:
              - <pkgA_slice-name>
              - ...

            # (req) The list of files, from the package, that this slice will install
            contents:
                </path/to/content>:
                </path/to/another/multiple*/content/**>:
                </path/to/moved/content>: {copy: /bin/original}
                </path/to/link>: {symlink: /bin/mybin}
                </path/to/new/dir>: {make: true}
                </path/to/file/with/text>: {text: "Some text"}
                </path/to/mutable/file/with/default/text>: {text: FIXME, mutable: true}
                </path/to/temporary/content>: {until: mutate}

>             # (opt) Mutation scripts, to allow for the reproduction of maintainer scripts, based on Starlark (https://github.com/canonical/starlark)

Reviewer: ```suggestion
            # (opt) Mutation scripts, to allow for the reproduction of maintainer scripts,
            # based on Starlark (https://github.com/canonical/starlark)
```

            mutate: |
                ...

Find the dependencies of your package
-------------------------------------

> So, the first step is to find the dependencies of the package for which you want to create a new slice definition (``openssl`` in this guide). You can easily do it by:

Reviewer: ```suggestion
So, the first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:
```

.. code-block:: sh

    $ apt show openssl

> .. code-block::

Reviewer: ```suggestion
.. code-block:: text
```

    package: openssl
    Version: 3.0.2-0ubuntu1.7
    Origin: Ubuntu
    Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
    Original-Maintainer: Debian OpenSSL Team <pkg-openssl-devel@alioth-lists.debian.net>
    Bugs: https://bugs.launchpad.net/ubuntu/+filebug
    Depends: libc6 (>= 2.34), libssl3 (>= 3.0.2-0ubuntu1.2)

> From the above output, you can confirm that ``openssl`` **depends on** ``libc6`` and ``libssl3``. So when creating your slice definition file for OpenSSL, you'll need to remember to include those packages slices as a dependency as well, whenever needed. Let's do that in the following section.

Reviewer: ```suggestion
From the above output, you can confirm that ``openssl`` **depends on** ``libc6`` and ``libssl3``. So when creating your slice definitions file for OpenSSL, you'll need to remember to include those packages' slices as a dependency as well, whenever needed. Let's do that in the following section.
```

Create your slice definition
----------------------------

You now have everything needed to actually define the OpenSSL slice that will install the content you're looking to have in the final ROCK.

> Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins** ?  Let's go for it:

Reviewer: ```suggestion
Since you're looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**?  Let's go for it:
```

3. **What is your slice name?** It should be called ``bins``
4. **What contents do you need from the OpenSSL package?** Just the binaries - ``/usr/bin/c_rehash`` and ``/usr/bin/openssl``
5. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``
    - **Do these two packages have slice definitions files upstream?** Yes, there's already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_
        - If these dependencies were not present in the upstream Chisel release, you'd also need to create their corresponding slice definitions
    - **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` itself.

With all this figured out, you can now create a new YAML file named ``openssl.yaml``, with the following content:

.. code-block:: yaml

>     package: openssl

Reviewer: let's fix the indentation:

```
    package: openssl
    slices:
        bins:
            essential:
              - libc6_libs
              - libc6_config
              - libssl3_libs
              - openssl_config
            contents:
                /usr/bin/c_rehash:
                /usr/bin/openssl:

        config:
            contents:
                /etc/ssl/private/:
                /etc/ssl/openssl.cnf:
                /usr/lib/ssl/certs:
                /usr/lib/ssl/openssl.cnf:
                /usr/lib/ssl/private:
```

How to create a package slice for Chisel
****************************************

If your package doesn't yet have the slice definitions you actually need to **create your own slice definition** (which you can, later on, propose to be merged upstream for everyone else to use :ref:`publishslice` ).

**Let's assume you are trying to create a slice definition for installing the OpenSSL binary into your ROCK!**

Make sure the slice definition doesn't exist yet
------------------------------------------------

> To avoid re-creating a slice that already exist, check the following to see if something that fits your needs already exists:

Reviewer 2: ```suggestion
To avoid re-creating a slice, check the following to see if something that fits your needs already exists:
```
This way we avoid repeating "exists"

> To avoid re-creating a slice that already exist, check the following to see if something that fits your needs already exists:

Reviewer: ```suggestion
To avoid re-creating a slice that already exists, check the following to see if something that fits your needs already exists:
```

#. Look into the upstream `chisel-releases <https://github.com/canonical/chisel-releases/tree/ubuntu-22.04/slices>`_ repository
#. Switch to the branch corresponding to the desired Ubuntu release for your ROCK

> #. Search your package name through the list of slice definition files

Reviewer: ```suggestion
#. Search your package name within the list of slice definitions files
```

   * if you find it, open it and try to find a slice name containing the bits and pieces you need from that package

> Structure of a slice definitions file

Reviewer 2: This section feels a bit disconnected from the rest of the guide: it reads more like a reference. Is it indispensable here?

Reviewer: could we keep this one here, for now, and add an issue for me to move it to the Chisel reference? (I can't move it now, because we don't have a Chisel reference, and I don't think it should be in the Rockcraft documentation)

Reviewer: created https://github.com/canonical/rockcraft/issues/157

------------------------------------

> There is **one slice definitions file** for each Ubuntu package. All of the slice definitions files follow the same structure:

Reviewer 2: ```suggestion
There can be only **one slice definitions file** for each Ubuntu package. All of the slice definitions files follow the same structure:
```
My reasoning here is that saying "There is one file for each package" can make people think that every package already has a slice file, which is not true. I believe the intent was to express that each package only has one slice file, right?

Author: For me you are right, i have updated it

Reviewer: :+1:

    # (req) Name of the package.
    # The slice definition file should be named accordingly (eg. "openssl.yaml")

    package: <package-name>

    # (req) List of slices
    slices:

        # (req) Name of the slice
        <slice-name>:

            # (opt) Optional list of slices that this slice depends on
            essential:
              - <pkgA_slice-name>
              - ...

            # (req) The list of files, from the package, that this slice will install
            contents:
                </path/to/content>:
                </path/to/another/multiple*/content/**>:
                </path/to/moved/content>: {copy: /bin/original}
                </path/to/link>: {symlink: /bin/mybin}
                </path/to/new/dir>: {make: true}
                </path/to/file/with/text>: {text: "Some text"}
                </path/to/mutable/file/with/default/text>: {text: FIXME, mutable: true}
                </path/to/temporary/content>: {until: mutate}

            # (opt) Mutation scripts, to allow for the reproduction of maintainer scripts,
            # based on Starlark (https://github.com/canonical/starlark)
            mutate: |
                ...

> Find the dependencies of your package

Reviewer 2: I like this section, but to me one of the hardest parts of Diátaxis is trying to avoid explaining to much; for how-tos one [should not explain concepts](https://diataxis.fr/how-to-guides/#don-t-explain-concepts) if possible so I actually don't know if this is too much explanation or not! Thoughts?
Maybe we should ask someone on the Documentation team for help? Can either of you attend the Documentation Office Hours today?

Author: When are the Documentation office hours ?

Reviewer 2: An hour from this comment; this is the meeting: https://meet.google.com/vpj-sanz-mrb

Reviewer 2: I asked because I can't make it to today's; on second thought it might be better if we finish reviewing this batch and get it merged, consolidate our questions and bring them to Doc office hours next week

Reviewer: IMO, this specific section is fine. It doesn't explain stuff, but rather guide you through a process. The one above though, with the slice schema, indeed feels out of place here. I'll comment there

-------------------------------------

> The first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). It's done like this:

Reviewer 2: ```suggestion
The first step is to find the dependencies of the package for which you want to create a new slice definition (``openssl`` in this guide). It's done like this:
```

> So, the first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:

Reviewer 3: ```suggestion
The first step is to find the dependencies of the package for which you want to create a new slice definitions (``openssl`` in this guide). You can easily do it by:
```

.. code-block:: sh

    $ apt show openssl

.. code-block:: text

    package: openssl
    Version: 3.0.2-0ubuntu1.7
    Origin: Ubuntu
    Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
    Original-Maintainer: Debian OpenSSL Team <pkg-openssl-devel@alioth-lists.debian.net>
    Bugs: https://bugs.launchpad.net/ubuntu/+filebug
    Depends: libc6 (>= 2.34), libssl3 (>= 3.0.2-0ubuntu1.2)

From the above output, you can confirm that ``openssl`` **depends on** ``libc6`` and ``libssl3``. So when creating your slice definitions file for OpenSSL, you will need to remember to include those packages' slices as a dependency as well, whenever needed. Let's do that in the following section.

Create your slice definition
----------------------------

> You now have everything needed to actually define the OpenSSL slice that will install the content you are looking to have in the final ROCK.Since you are looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**? Let's go for it:

Reviewer 2: ```suggestion
You now have everything needed to actually define the OpenSSL slice that will install the content you are looking to have in the final ROCK. Since you are looking to install just the OpenSSL binaries from the ``openssl`` package, what about naming this new slice **bins**? Let's go for it:
```

#. **What is the name of your slice definitions file?** It is a YAML file called ``openssl.yaml``
#. **What package name should be defined inside this file?** The package name is ``openssl``
#. **What is your slice name?** It should be called ``bins``
#. **What contents do you need from the OpenSSL package?** Just the binaries - ``/usr/bin/c_rehash`` and ``/usr/bin/openssl``
#. **Does your slice depend on any other package slice?** Yes, OpenSSL depends on ``libc6`` and ``libssl3``

>    * **Do these two packages have slice definitions files upstream?** Yes, there is already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_ . If these dependencies were not present in the upstream Chisel release, you will also need to create their corresponding slice definitions

Reviewer 2: ```suggestion
   * **Do these two packages have slice definitions files upstream?** Yes, there is already a slice definitions file for `libc6 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libc6.yaml>`_ and another one for `libssl3 <https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/libssl3.yaml>`_ . If these dependencies were not present in the upstream Chisel release, you would also need to create their corresponding slice definitions
```

>    * **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` itself.

Reviewer 2: ```suggestion
   * **Which slices do you depend on then?** Since you only want the OpenSSL binaries, you might only need the libraries from ``libc6`` and ``libssl3``, as well as the configuration files from ``libc6`` and ``openssl`` themselves.
```

Create a new YAML file named ``openssl.yaml``, with the following content:

.. code-block:: yaml

    package: openssl
    slices:
        bins:
            essential:
              - libc6_libs
              - libc6_config
              - libssl3_libs
              - openssl_config
            contents:
                /usr/bin/c_rehash:
                /usr/bin/openssl:

        config:
            contents:
                /etc/ssl/private/:
                /etc/ssl/openssl.cnf:
                /usr/lib/ssl/certs:
                /usr/lib/ssl/openssl.cnf:
                /usr/lib/ssl/private:

> Notice the unforeseen new slice ``config``. Because your OpenSSL binaries depend on the OpenSSL configuration files, and those were not yet present anywhere in the Chisel releases upstream, you also need to create that slice! You may also ask **"why not put those configuration files inside the ``bins`` slice"**? You could! But we recommend, as a best practice, to separate and group contents according to their nature, as you may tomorrow need to create a new slice definition that only needs the OpenSSL configurations and not the binaries.

Reviewer 2: I think something that we need to address in documentation somehow in the future (not on this PR necessarily) is that the jump from "I just need the bins" to "I also need this bunch of config" feels a bit jarring. I know that someone (was it you @cjdcordeiro?) had to "suffer" when creating these slices, but on this how-to the new requirement feels a bit like it came out of nowhere.

In practice, that's probably going to be one of the harder parts when creating new slices, right? Is that something that would merit a separate how-to/troubleshooting/whatever section?

Reviewer: This is really open to developer discretion. Like for DEBs, how do you say what pkgs your DEB depends on? First, the definition must exist, for one to be able to tell what does the slice depends on. In this case, that's the definition we are working on here...it does require some in-depth knowledge about the OpenSSL inner workings

How to create a package slice for Chisel
****************************************

> You want to build a chiselled ROCK, but Chisel cannot find the exact bits and pieces of the packages you need to install! This means that said package doesn't yet have the slice definitions you're looking for. In these cases, your best move is to actually **create your own slice definition** (which you can, later on, propose to be merged upstream for everyone else to use).

Reviewer 3: Can we use conditional imperatives here https://diataxis.fr/how-to-guides/#the-language-of-how-to-guides ?

If we move to the imperative part, I would remove "You want to build a chiselled ROCK, but Chisel cannot find the exact bits and pieces of the packages you need to install!"

Side comment, can we standarize on not relying on line wrapping? It makes it easier to review and spatially read the source

"which you can, later on, propose to be merged upstream for everyone else to use" should eventually point to a howto (or where upstream is)

Author: I have modified the ref publish-slice; it refers to the publish slice PR. Is it good now?

Author: @sergiusens the PR looks good to you, or does it need other changes?

**Let's assume you are trying to create a slice definition for installing the OpenSSL binary into your ROCK!**

> Make sure the slice definition doesn't exist yet

Reviewer 3: we should not use contractions https://docs.ubuntu.com/styleguide/en

Reviewer 3: well, doesn't is allowed....

Author: Can i resolve this conversation?

-------------------------------------------------------------

> You obviously want to avoid re-doing existing work, so before even considering creating your own slice definition, please make sure there isn't already something fitting your needs. **How** ?

Reviewer 3: We should not assume something is obvious, switch to the imperative form, we should not be saying please

```suggestion
To avoid re-creating a slice that already exist, check the following to see if something that fits your needs already exists:
```
